tornado/websocket.py

```diff
@@ -1082,10 +1093,17 @@ def write_message(
             opcode = 0x2
         else:
             opcode = 0x1
-        message = tornado.escape.utf8(message)
-        assert isinstance(message, bytes)
-        self._message_bytes_out += len(message)
+
         flags = 0
+
+        if isinstance(message, list):
```

Comment:
If one websocket message actually consists of multiple parts, it is desirable to give all of them to this function as an array. Instead of that, Bokeh splits one big logical message to small ones in its exchange protocol. This is not optimal, and new versions might take advantage of new Tornado feature.

Writing multiple websocket messages using one function call will not help against memory copying though.

Reply:
@bdarnell Currently Bokeh splits up messages like this:
```
    [
        # these are required
        b'{header}',        # serialized header dict
        b'{metadata}',      # serialized metadata dict
        b'{content},        # serialized content dict

        # these are optional, and come in pairs; header contains num_buffers
        b'{buf_header}',    # serialized buffer header dict
        b'array'            # raw buffer payload data
        ...
    ]
```
The header fragment will have the form:
```
    header = {
        # these are required
        'msgid'       : <str> # a unique id for the message
        'msgtype'     : <str> # a message type, e.g. 'ACK', 'PATCH-DOC', etc

        # these are optional
        'num_buffers' : <int> # the number of additional buffers, if any
    }
```

The `metadata` fragment has never been used. We could remove it for 2.0. The `content` and buffer arrays are potentially large, which is why I made the protocol split them out, so that the potentially large copies/concats to assemble a single message would be avoided. 

If it is better to concat smaller fragments in order to send a single message rather then multiple, then an optimization we could potentially make on the Bokeh side would be to concat the main header and any buffer headers up front, and just leave content or array fragments separate. Do you think that is advisable?

Reply:
@bdarnell @socketpair any feedback on these comments? Is there something I can elaborate on?

Reply:
Well. I think, there are many three ways how to make Bokeh faster:
1. Change websocket internal protocol - i.e. move from sending small messages to one contiguous message. Splitting to multiple messages had been made in order to eliminate memory copying in Tornado. Now it is not the case.
2. Think about `yield` at left side of every websocket send. If present -- it's correct -- it catch errors, waits fro send completion, stops server from sending next message if previous one has not been sent. Without it -- memory consumption may be raised, but latency should be much lower + throughput is higher (since no pauses between messages).
3. Switch to aiohttp :). Tornado is way slower than aiohttp in all cases I know, from websocket handling to coroutines. Also it has uvloop. Possibly only some of the requests can be moved to aiohttp. I don't think the move will be difficult. Yes, you have to change every source file by hands, but changes are simple and template-like.